pkg/controller/periodic/periodic.go

```diff
@@ -24,101 +23,97 @@ type Configurator interface {
 	}
 	return sources
 }
 
-// sync gathers data from the cluster periodically
-func (c *Controller) sync(name string) error {
+func (c *Controller) Run(stopCh <-chan struct{}, initialDelay time.Duration) {
+	defer utilruntime.HandleCrash()
+	defer klog.Info("Shutting down")
+	c.initialDelay = initialDelay
+
+	go wait.Until(func() { c.periodicTrigger(stopCh) }, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) Gather() {
+	for name := range c.gatherers {
+		start := time.Now()
+		err := c.runGatherer(name)
+		if err == nil {
+			klog.V(4).Infof("Periodic gather %s completed in %s", name, time.Since(start).Truncate(time.Millisecond))
+			c.statuses[name].UpdateStatus(controllerstatus.Summary{Healthy: true})
+			return
+		}
+		utilruntime.HandleError(fmt.Errorf("%v failed after %s with: %v", name, time.Since(start).Truncate(time.Millisecond), err))
+		c.statuses[name].UpdateStatus(controllerstatus.Summary{Reason: "PeriodicGatherFailed", Message: fmt.Sprintf("Source %s could not be retrieved: %v", name, err)})
+	}
+}
+
+func (c *Controller) runGatherer(name string) error {
 	gatherer, ok := c.gatherers[name]
 	if !ok {
 		klog.V(2).Infof("No such gatherer %s", name)
 		return nil
 	}
-	ctx, cancel := context.WithTimeout(context.Background(), c.config.Config().Interval/2)
+	ctx, cancel := context.WithTimeout(context.Background(), c.configurator.Config().Interval/2)
 	defer cancel()
 	defer func() {
 		if err := c.recorder.Flush(ctx); err != nil {
 			klog.Errorf("Unable to flush recorder: %v", err)
 		}
 	}()
 	klog.V(4).Infof("Running %s", name)
-	return gatherer.Gather(ctx, c.config.Config().Gather, c.recorder)
-}
-
-// Run starts gathering with initialDelay
-func (c *Controller) Run(workers int, stopCh <-chan struct{}, initialDelay time.Duration) {
-	defer utilruntime.HandleCrash()
-	defer c.queue.ShutDown()
-	defer klog.Info("Shutting down")
-
-	// start watching for version changes
-	go wait.Until(func() { c.periodicTrigger(stopCh) }, time.Second, stopCh)
-
-	for i := 0; i < workers; i++ {
-		go wait.Until(func() {
-			if initialDelay > 0 {
-				select {
-				case <-stopCh:
-					return
-				case <-time.After(initialDelay):
-					c.runWorker()
-				}
-			}
-			c.runWorker()
-		}, time.Second, stopCh)
-	}
-
-	// seed the queue
-	c.Gather()
-
-	<-stopCh
-}
-
-func (c *Controller) Gather() {
-	for name := range c.gatherers {
-		c.queue.Add(name)
-	}
+	return gatherer.Gather(ctx, c.configurator.Config().Gather, c.recorder)
 }
 
 func (c *Controller) periodicTrigger(stopCh <-chan struct{}) {
-	configCh, closeFn := c.config.ConfigChanged()
+	configCh, closeFn := c.configurator.ConfigChanged()
 	defer closeFn()
 
-	interval := c.config.Config().Interval
+	if c.initialDelay > 0 {
+		select {
+		case <-stopCh:
+			return
+		case <-time.After(c.initialDelay):
+			c.initialDelay = 0
+			c.Gather()
+		}
+	}
+
+	interval := c.configurator.Config().Interval
 	expireCh := time.After(wait.Jitter(interval, 0.5))
```

Comment:
Do we still need the `expireCh`?

Reply:
no idea, its something that was in the original and I didn't dare to remove it
will check what will happen if I remove it